kernel/src/engine/default/json.rs

```diff
@@ -154,7 +154,8 @@ impl<E: TaskExecutor> JsonHandler for DefaultJsonHandler<E> {
             }
         });
 
-        Ok(Box::new(receiver.into_iter().map(|rbr| {
+        let iter = std::iter::from_fn(move || receiver.blocking_recv());
```

Comment:
this `blocking_recv()` call is the source of the `"Cannot block the current thread from within a runtime."` issue.

Some of our unit tests are marked as `#[tokio:test]` which puts the test into an async context, and so when this code gets run we the runtime notices we're trying to call `blocking_recv` and panics.

We could do something where we try and detect if we're in a runtime with something like [`Handle::try_current`](https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.try_current), and then spawn a normal async `recv().await` if we are. This has a couple of downsides:

1. More code paths, and generally the unit test path will not be the commonly used one
2. Reliance on tokio. At this point in the code we want to _only_ have an `executor` and not be dependent directly on `tokio` to enable engines to possibly use a different runtime if they provide an `executor` implementation. We'd need to add a `try_current` method to the api of executor or something to abstract that.

I think for now I'd maybe just wrap the old channel in a `Mutex` and we can revisit this specific thing in the future.